samanthas_telegram_bot/api_clients/smalltalk/smalltalk_client.py

```diff
@@ -0,0 +1,273 @@
+"""Functions for interaction with SmallTalk oral test service."""
+import asyncio
+import logging
+import os
+import typing
+
+import httpx
+from dotenv import load_dotenv
+from telegram import Update
+from telegram.constants import ParseMode
+
+from samanthas_telegram_bot.api_clients.auxil.constants import (
+    MAX_ATTEMPTS_TO_GET_DATA_FROM_API,
+    SMALLTALK_TIMEOUT_IN_SECS_BETWEEN_API_REQUEST_ATTEMPTS,
+    SMALLTALK_URL_GET_RESULTS,
+    SMALLTALK_URL_GET_TEST,
+    DataDict,
+)
+from samanthas_telegram_bot.api_clients.auxil.enums import LoggingLevel
+from samanthas_telegram_bot.api_clients.auxil.models import NotificationParams
+from samanthas_telegram_bot.api_clients.base.base_api_client import BaseApiClient
+from samanthas_telegram_bot.api_clients.base.exceptions import BaseApiClientError
+from samanthas_telegram_bot.api_clients.smalltalk.exceptions import (
+    SmallTalkClientError,
+    SmallTalkJSONParsingError,
+    SmallTalkLogicError,
+    SmallTalkRequestError,
+)
+from samanthas_telegram_bot.auxil.escape_for_markdown import escape_for_markdown
+from samanthas_telegram_bot.auxil.log_and_notify import logs
+from samanthas_telegram_bot.data_structures.constants import ALL_LEVELS
+from samanthas_telegram_bot.data_structures.context_types import CUSTOM_CONTEXT_TYPES
+from samanthas_telegram_bot.data_structures.enums import SmalltalkTestStatus
+from samanthas_telegram_bot.data_structures.models import SmalltalkResult
+
+load_dotenv()
+logger = logging.getLogger(__name__)
+
+HEADERS = {"Authorization": f"Bearer {os.environ.get('SMALLTALK_TOKEN')}"}
+ORAL_TEST_ID = os.environ.get("SMALLTALK_TEST_ID")
+
+
+class SmallTalkClient(BaseApiClient):
+    @classmethod
+    async def send_user_data_get_test(
+        cls,
+        update: Update,
+        context: CUSTOM_CONTEXT_TYPES,
+    ) -> tuple[str | None, str | None]:
+        """Gets SmallTalk interview ID and test URL."""
+        user_data = context.user_data
+
+        try:
+            _, data = await cls.post(
+                update=update,
+                context=context,
+                url=SMALLTALK_URL_GET_TEST,
+                headers=HEADERS,
+                json_data={
+                    "test_id": ORAL_TEST_ID,
+                    "first_name": user_data.first_name,
+                    "last_name": user_data.last_name,
+                    "email": user_data.email,
+                },  # TODO possibly webhook
+                notification_params_for_status_code={
+                    httpx.codes.OK: NotificationParams(
+                        "Received data that should contain the link to SmallTalk test"
+                    ),
+                },
+            )
+        except BaseApiClientError as err:
+            raise SmallTalkRequestError(
+                "Failed to get data with test link from SmallTalk"
+            ) from err
+
+        try:
+            url = data.get("test_link")
+        except AttributeError as err:
```

Comment:
How AttributeError could occur here? 
`data` is not always a dict? why?

Reply:
Hm, I think it's some old code left here where I got a response and didn't know whether data was `None` or not. I reworked all of that since with exceptions in (hopefully) the right places, so this `try` and the next one are no longer needed. Thanks.